// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
syntax = "proto2";

package koladata.s11n;

import "arolla/serialization_base/base.proto";

option cc_enable_arenas = true;

message KodaV1Proto {
  extend arolla.serialization_base.ValueProto {
    optional KodaV1Proto extension = 329924649;
  }

  message ObjectIdProto {
    optional uint64 hi = 1;
    optional uint64 lo = 2;
  }

  message DataItemProto {
    oneof value {
      bool missing = 1;
      ObjectIdProto object_id = 2;
      int32 i32 = 3;
      int64 i64 = 4;
      float f32 = 5;
      double f64 = 6;
      bool boolean = 7;
      bool unit = 8;
      string text = 9;
      bytes bytes_ = 10;
      int32 dtype = 11;
      bool expr_quote = 12;  // stored in ValueProto.input_value_indices[0]
    }
  }

  message DataItemVectorProto {
    repeated DataItemProto values = 1;
  }

  message DataSliceImplProto {
    oneof value {
      DataItemVectorProto data_item_vector = 1;
      // Note: later we can add optimized representations e.g.
      // serialized DenseArrays of each type linked via input_value_indices.
    }
  }

  // Represents values of some attribute (attribute name not included) for
  // a sequential range of ObjectIds.
  // `values` can be either DataItem (in this case first_object_id is
  // the ObjectId for the item), or DataSliceImpl. In case of DataSliceImpl
  // first_object_id determines allocation and a starting offset in
  // the allocation. Currently offset is always zero, but potentially we can
  // split a huge (> 2GiB) allocation into several chunks and serialize
  // separately.
  message AttrChunkProto {
    optional ObjectIdProto first_object_id = 1;
    // index of DataSliceImpl/DataItem in DataBag's
    // ValueProto.input_value_indices
    optional int32 values_subindex = 2;
  }

  message AttrProto {
    optional string name = 1;
    repeated AttrChunkProto chunks = 2;
  }

  message DictProto {
    optional ObjectIdProto dict_id = 1;
    // index of keys DataSliceImpl (not DataItem) in
    // DataBag's ValueProto.input_value_indices
    optional int32 keys_subindex = 2;
    // index of values DataSliceImpl (not DataItem) in
    // DataBag's ValueProto.input_value_indices
    optional int32 values_subindex = 3;
  }

  message ListProto {
    optional ObjectIdProto list_id = 1;
    // index of values DataSliceImpl in DataBag's ValueProto.input_value_indices
    optional int32 values_subindex = 2;
  }

  // ValueProto.input_value_indices:
  //     First fallback_count indices are links to fallbacks (DataBagProto)
  //     of this DataBag. Other indices are links to
  //     DataItemProto/DataSliceImplProto referenced in attrs, dicts, lists.
  message DataBagProto {
    optional int32 fallback_count = 1;
    repeated AttrProto attrs = 2;
    repeated DictProto dicts = 3;
    repeated ListProto lists = 4;
  }

  oneof value {
    // Represents koladata::expr::LiteralOperator
    //
    // ValueProto.input_value_indices[0]
    //   -- Wrapped TypedValue.
    bool literal_operator = 1;
    bool ellipsis_value = 2;
    DataItemProto internal_data_item_value = 3;
    DataSliceImplProto data_slice_impl_value = 4;

    DataBagProto data_bag_value = 5;

    // ValueProto.input_value_indices[0]: DataSliceImpl or DataItem
    // ValueProto.input_value_indices[1]: JaggedShape
    // ValueProto.input_value_indices[2]: Schema (DataItem)
    // ValueProto.input_value_indices[3]: (optional) DataBag
    bool data_slice_value = 6;

    // Represents koladata::DataSlice's QType
    bool data_slice_qtype = 10;

    bool ellipsis_qtype = 11;
    bool data_bag_qtype = 12;
    bool non_deterministic_token_qtype = 13;
  }
}
